conda/__init__.py

```diff
@@ -148,3 +139,209 @@ def _default(self, obj):
+
+        return self
+
+    @staticmethod
+    def _get_module(stack: int) -> tuple[ModuleType, str]:
+        import inspect  # expensive
+
+        try:
+            frame = inspect.stack()[2 + stack]
+            module = inspect.getmodule(frame[0])
+            return (module, module.__name__)
+        except (IndexError, AttributeError):
+            raise RuntimeError("unable to determine the calling module") from None
+
+    @classmethod
+    def module(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        *,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation function for modules."""
+        cls.topic(
+            deprecate_in=deprecate_in,
+            remove_in=remove_in,
+            topic=cls._get_module(stack)[1],
+            addendum=addendum,
+            stack=stack + 2,
+        )
+
+    @classmethod
+    def constant(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        constant: str,
+        value: Any,
+        *,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation function for module constant (global)."""
+        # detect calling module
+        module, fullname = cls._get_module(stack)
+
+        self = cls(deprecate_in=deprecate_in, remove_in=remove_in, addendum=addendum, stack=stack)
+
+        # alert developer that it's time to remove something
+        if not self._category:
+            raise RuntimeError(self._message(f"{fullname}.{constant}"))
+
+        # patch module level __getattr__ to alert user that it's time to remove something
+        super_getattr = getattr(module, "__getattr__", None)
+
+        def __getattr__(name: str) -> Any:
+            if name == constant:
+                warnings.warn(
+                    self._message(f"{fullname}.{name}"), self._category, stacklevel=2 + self._stack
+                )
+                return value
+
+            if super_getattr:
+                return super_getattr(name)
+
+            raise AttributeError(f"module '{fullname}' has no attribute '{name}'")
+
+        module.__getattr__ = __getattr__
+
+    @classmethod
+    def topic(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        *,
+        topic: str,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation function for a topic."""
+        self = cls(deprecate_in=deprecate_in, remove_in=remove_in, addendum=addendum, stack=stack)
+
+        # alert developer that it's time to remove something
+        if not self._category:
+            raise RuntimeError(self._message(topic))
+
+        # alert user that it's time to remove something
+        warnings.warn(self._message(topic), self._category, stacklevel=2 + self._stack)
+
+    @classmethod
+    def _factory(cls, version: str) -> _deprecated:
+        return type("_deprecated", (cls,), {"_version": cls._parse_version(version)})
```

Comment:
Assuming we agree to allow other developers of conda related projects to make use of this, I think we should use a class to clearly instantiate the `decorated` helper, e.g. in a `conda_acme/__init__.py`:

```python
from conda import deprecations

__version__ = "23.1.0"

# A deprecation helper using the current version
deprecated = deprecations.Handler(__version__)

```

where `Handler` is a regular class with a number of instance methods for each use case (argument, constant etc)

Reply:
This was the intended use case I had envisioned for the `_factory` method, renaming it to `Handler` confuses me.

Is the intention to push `__init__` and `__call__` down a level? So instead of:

```
@deprecated(...)
def foo(...):
    ...
```

We do:

```
@deprecated.function(...)
def foo(...):
    ...
```

Reply:
My concerns is that the class does too much for its own sake, wrapping various functionality into one class (a generic factory, a few specific factories per decorated type, a general helper to retain state, a decorator).  The "handler" name was a suggestion to encourage being clear about it being a complex class and decoupling some of its parts.

Other than that, it's also trying to be "private" (lots of underscores) where [we know](https://www.hyrumslaw.com/) and want to encourage that it'll be used long-term, so why even try to hide the implementation and not optimize for ease of understanding?

My guess is that you're trying to make this code as small as possible, but it's impacting the readability and complexity of the code without an upside.

Reply:
all good points, I def wanted to make this as small as possible while still covering our use cases (conflicting objectives...)